### ISSUE TITLE
use_gobject_class_install_properties fails with g_object_class_override_property()

### ISSUE THREAD

**Issue by benjamin-otte** [bilelmoussaoui/gobject-lint#102]
Afaik there's no documented way to use `g_object_class_override_property()` with `g_object_class_install_properties()` so the rule should probably not even flag classes that use `g_object_class_override_property()`.

But the fixes it produces are definitely wrong because the overridden pspecs do have enum values but they don't end up in the pspec array. So the call to `g_object_class_install_properties()` contains NULL values.

**Comment by benjamin-otte** [comment]
I was thinking about replacing `g_object_class_override_property()` with [g_param_spec_override()](https://docs.gtk.org/gobject/func.param_spec_override.html) because that would also allow using `g_object_notify_by_pspec()` but I'm not sure what GObject developers think about that.

And another approach would be to rewrite the enum to have overwritten properties at the end, kinda like this [gtkviewport.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/95452f7e/gtk/gtkviewport.c#L86)

```c
enum {
  PROP_0,
  PROP_SCROLL_TO_FOCUS,
  PROP_CHILD,
  N_REAL_PROPS,

  PROP_HADJUSTMENT = N_REAL_PROPS,
  PROP_VADJUSTMENT,
  PROP_HSCROLL_POLICY,
  PROP_VSCROLL_POLICY,
  N_PROPS
};

GParamSpec pspecs[N_REAL_PROPS] = { NULL, };
```

**Comment by bilelmoussaoui** [comment]
> And another approach would be to rewrite the enum to have overwritten properties at the end, kinda like this [gtkviewport.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/95452f7e/gtk/gtkviewport.c#L86)
> 
> enum {
>   PROP_0,
>   PROP_SCROLL_TO_FOCUS,
>   PROP_CHILD,
>   N_REAL_PROPS,
> 
>   PROP_HADJUSTMENT = N_REAL_PROPS,
>   PROP_VADJUSTMENT,
>   PROP_HSCROLL_POLICY,
>   PROP_VSCROLL_POLICY,
>   N_PROPS
> };
> 
> GParamSpec pspecs[N_REAL_PROPS] = { NULL, };


That is the pattern I found in various projects like Mutter. Not sure what is the better pattern to suggest from the rule level.

**Comment by benjamin-otte** [comment]
I just found https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkcolumnview.c#L812-825 which does:

```c
  /* GtkScrollable implementation */
  iface = g_type_default_interface_peek (GTK_TYPE_SCROLLABLE);
  properties[PROP_HADJUSTMENT] =
      g_param_spec_override ("hadjustment",
                             g_object_interface_find_property (iface, "hadjustment"));
```

That way the property is available for `g_object_class_install_properties()` and (and also `g_object_notify_by_pspec()`, see #103). But it requires knowing which interface to query the property from.

(It's great that this is my code but I don't remember anything about coming up with that.)

**Comment by bilelmoussaoui** [comment]
> I just found https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkcolumnview.c#L812-825 which does:
> 
>   /* GtkScrollable implementation */
>   iface = g_type_default_interface_peek (GTK_TYPE_SCROLLABLE);
>   properties[PROP_HADJUSTMENT] =
>       g_param_spec_override ("hadjustment",
>                              g_object_interface_find_property (iface, "hadjustment"));
> 
> That way the property is available for `g_object_class_install_properties()` and (and also `g_object_notify_by_pspec()`, see [#103](https://github.com/bilelmoussaoui/gobject-linter/issues/103)). But it requires knowing which interface to query the property from.
> 
> (It's great that this is my code but I don't remember anything about coming up with that.)

That looks neat but as you said, we don't have a way to figure what from which iface that property comes from. It should be doable for local ifaces but that is not always the case. e.g, an app implementing Scrollable iface

**Comment by bilelmoussaoui** [comment]
> I just found https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkcolumnview.c#L812-825 which does:
> 
>   /* GtkScrollable implementation */
>   iface = g_type_default_interface_peek (GTK_TYPE_SCROLLABLE);
>   properties[PROP_HADJUSTMENT] =
>       g_param_spec_override ("hadjustment",
>                              g_object_interface_find_property (iface, "hadjustment"));
> 
> That way the property is available for `g_object_class_install_properties()` and (and also `g_object_notify_by_pspec()`, see [#103](https://github.com/bilelmoussaoui/gobject-linter/issues/103)). But it requires knowing which interface to query the property from.
> 
> (It's great that this is my code but I don't remember anything about coming up with that.)

This is now automatically handled for you. 

Although, we can't do that if the iface is defined in a separate project like I mentioned above. For that to work, we would need to consume GIR data as well which is something I am planning to work on later.

I guess both this and #103 can be closed now

**Comment by benjamin-otte** [comment]
Just reran `g_object_class_install_properties --fix` on the GTK sources (on [94e6a1c4](https://gitlab.gnome.org/GNOME/gtk/-/tree/94e6a1c4) to be exact) and [gtkcellrenderer.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/94e6a1c4/gtk/deprecated/gtkcellrenderer.c) was the only file that it didn't fix.

It also didn't mark it as a violation - neither before nor after I had done the `--fix` run.

**Comment by bilelmoussaoui** [comment]
> Just reran `g_object_class_install_properties --fix` on the GTK sources (on [94e6a1c4](https://gitlab.gnome.org/GNOME/gtk/-/tree/94e6a1c4) to be exact) and [gtkcellrenderer.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/94e6a1c4/gtk/deprecated/gtkcellrenderer.c) was the only file that it didn't fix.
> 
> It also didn't mark it as a violation - neither before nor after I had done the `--fix` run.

Because that file doesn't use any G_DEFINE macro as at parse time we associate properties to their corresponding GObject. How common it is to have gobjects using `g_type_register_static`? 

I guess we can do the other direction, start from class_init and skip the finding gobject part, but we have various helpers in that direction that we wouldn't be able to re-use anymore.

**Comment by benjamin-otte** [comment]
Oh hm. I know we use `register_static()` with certain very base classes like `GtkWidget` and apparently `GtkCellRenderer` - usually because we want access to the `GTypeInfo`.

`GtkWidget` uses it for the `base_init` function so it can clear some class structs that are per-class-instance, like the GBytes from [gtk_widget_class_set_template()](https://docs.gtk.org/gtk4/class_method.Widget.set_template.html).

Some classes use it because they need the 2nd argument in their `init()` function (`GtkListBase` uses [this hack](https://gitlab.gnome.org/GNOME/gtk/-/blob/4.22.0/gtk/gtklistbase.c#L107)).

TL;DR: I would expect some base classes or classes with special features to use it, so probably >10 cases but <100 cases across all the GObject using code. Obviously glib and GTK, most likely GStreamer, and of course [Gimp uses it](https://gitlab.gnome.org/search?group_id=8&nav_source=navbar&page=3&project_id=1848&repository_ref=master&search=g_type_register_static&search_code=true), but most of those are probably code from pre-G_DEFINE_TYPE days.

**Comment by bilelmoussaoui** [comment]
> Oh hm. I know we use `register_static()` with certain very base classes like `GtkWidget` and apparently `GtkCellRenderer` - usually because we want access to the `GTypeInfo`.
> 
> `GtkWidget` uses it for the `base_init` function so it can clear some class structs that are per-class-instance, like the GBytes from [gtk_widget_class_set_template()](https://docs.gtk.org/gtk4/class_method.Widget.set_template.html).
> 
> Some classes use it because they need the 2nd argument in their `init()` function (`GtkListBase` uses [this hack](https://gitlab.gnome.org/GNOME/gtk/-/blob/4.22.0/gtk/gtklistbase.c#L107)).
> 
> TL;DR: I would expect some base classes or classes with special features to use it, so probably >10 cases but <100 cases across all the GObject using code. Obviously glib and GTK, most likely GStreamer, and of course [Gimp uses it](https://gitlab.gnome.org/search?group_id=8&nav_source=navbar&page=3&project_id=1848&repository_ref=master&search=g_type_register_static&search_code=true), but most of those are probably code from pre-G_DEFINE_TYPE days.


We now parse register_static and the manual type declarations instead of G_DECLARE usages. That makes the tool capable of fixing must of the files. Although, things like _gtk_font_chooser_install_properties would make it produce wrong code for places that uses it, so needs manual handling in GTK case.

**Comment by bilelmoussaoui** [comment]
I suppose this and #103 are covered now?

**Comment by benjamin-otte** [comment]
This rule still warns about usage of `g_object_class_override_property()` - which is great for me because I'm replacing it with `g_param_spec_override()` calls manually, but I'm not sure if that's a good default behavior?

**Comment by bilelmoussaoui** [comment]
It does auto fix those as well, no?

**Comment by benjamin-otte** [comment]
Oh indeed, I wasn't on the latest git when I ran that.

Yeah, that's close this.

**Comment by benjamin-otte** [comment]
Looking at the result in more detail: Do you need to rewrite the props enum?

Because in some cases - [like the Mutter case you mentioned above](https://gitlab.gnome.org/GNOME/mutter/-/blob/ba18221a/src/backends/native/meta-seat-native.c#L45) - the `N_PROPS` is placed before the overridden properties.

**Comment by benjamin-otte** [comment]
Oh, also:

The fixing is not working with existing code that already uses install_properties() for non-overridden properties:

```c
g_object_class_override_property (gobject_class, PROP_INTERFACE, "interface-prop");

props[PROP_OWN] = g_param_spec_boolean ("my-own", NULL, NULL, FALSE, G_PARAM_READWRITE | G_PARAM_STATIC_NAME);

g_object_class_install_properties (gobject_class, N_PROPERTIES, props);
```

gets replaced as:

```c
props[PROP_INTERFACE] = g_param_spec_override ("interface-prop", g_type_default_interface_ref (MY_TYPE_INTERFACE), "interface-prop");

g_object_class_install_properties (gobject_class, N_PROPS, props);

props[PROP_OWN] = g_param_spec_boolean ("my-own", NULL, NULL, FALSE, G_PARAM_READWRITE | G_PARAM_STATIC_NAME);

g_object_class_install_properties (gobject_class, N_PROPS, props);
```

**Comment by bilelmoussaoui** [comment]
If you pull from main, and re-run the rule, it should fix all GTK violations and do proper fixes. I tested them locally :)

**Comment by blandford** [comment]
So, I just tried this from main with my code, and I'm getting it flagged in my code as an error. Code in question is:

```c
  g_object_class_override_property (object_class,
                                    PROP_ORIENTATION,
                                    "orientation");
```
Trying to fix it with `--fix` doesn't fix anything, and I can't tell from this long bug what the proposed final form should be for this. The suggestion is still the bad one though:

```shell
  error: Consider using g_object_class_install_properties() instead of 1 individual property installation calls
  rule: use_g_object_class_install_properties
```

**Comment by bilelmoussaoui** [comment]
with something like 
```
  /* GtkScrollable implementation */
  iface = g_type_default_interface_peek (GTK_TYPE_SCROLLABLE);
  properties[PROP_HADJUSTMENT] =
      g_param_spec_override ("hadjustment",
                             g_object_interface_find_property (iface, "hadjustment"));
```

we can't auto-fix because we don't know which iface that property comes from without integrating introspection data...

**Comment by benjamin-otte** [comment]
[gtk#403cb689](https://gitlab.gnome.org/GNOME/gtk/-/commit/403cb689) is half-manual half-automatic applying of this rule if you want an example.

**Comment by blandford** [comment]
That's helpful. Could we change the output of `--fix` to give a better hint as to what the expected resolution is. And maybe, is there any requirement to have `--fix` always be compileable: Could you produce the following code:

```c
  iface = g_type_default_interface_peek (G_TYPE_FIXME); // FIXME: replace this with the interface type that defines the "hadjustment" property 
  properties[PROP_HADJUSTMENT] =
      g_param_spec_override ("hadjustment",
                             g_object_interface_find_property (iface, "hadjustment"));
```

**Comment by bilelmoussaoui** [comment]
> That's helpful. Could we change the output of `--fix` to give a better hint as to what the expected resolution is. And maybe, is there any requirement to have `--fix` always be compileable: 

Eventually I want the tool to consume GIR data and be able to find the corresponding iface in all cases. 

> Could you produce the following code:
> 
>   iface = g_type_default_interface_peek (G_TYPE_FIXME); // FIXME: replace this with the interface type that defines the "hadjustment" property 
>   properties[PROP_HADJUSTMENT] =
>       g_param_spec_override ("hadjustment",
>                              g_object_interface_find_property (iface, "hadjustment"));

I think we should not generate code that wouldn't just work. But improving the message for override properties would make sense. Please open a separate issue about it =)

**Comment by benjamin-otte** [comment]
random thought: You could hardcode a few common interfaces and their properties like `GtkOrientable` to have a higher success rate at autofixing. In fact, the GTK MR above is just `GtkOrientable`, `GtkScrollable`, `GtkAccessible` and a few internal interfaces.
